AST reducer for validations

freshtonic opened the pull request:
This is essentially a functional-style implementation of the OO
"visitor" pattern.

freshtonic commented:
@joshprice @plasticine seeking feedback.

I'm happy with the small number of lines this is implemented in. I hope it's fairly easy to follow.

The idea is that you can call `ASTReducer.reduce(..., visitor, ...)` and have it return a an arbitrary data structure depending on the visitor you pass in.

I want to implement the `TypeInfo` visitor from `graphql-js` using this, next.

freshtonic commented on lib/graphql/lang/stack.ex:
```diff
@@ -0,0 +1,32 @@
+    %Stack{ stack | path: [node] ++ stack.path }
+  end
+
+  def pop(stack) do
```
I can actually get rid of `pop`. The nice thing about an immutable stack is that you never need to pop. The function just returns and the stack in the scope _is_ the popped version.

freshtonic commented on lib/graphql/lang/stack.ex:
```diff
@@ -0,0 +1,32 @@
+    %Stack{ stack | path: rest }
+  end
+
+  def peek(stack) do
```
I'm not using this yet, but I expect I will need to, so it will stay for now.

freshtonic commented on lib/graphql/lang/visitor_context.ex:
```diff
@@ -0,0 +1,6 @@
+
+defmodule GraphQL.Lang.VisitorContext do
```
Visitors can be provided with extra information during traversal. One such piece of information is the stack of nodes of the AST. In the future it will contain a `TypeInfo` map that would have been generated from an earlier pass.  Work in progress.

freshtonic commented on mix.exs:
```diff
@@ -16,6 +16,7 @@ defmodule GraphQL.Mixfile do
      homepage_url: @repo_url,
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
+     consolidate_protocols: Mix.env == :prod,
```
You need this if you defined a protocol implementation in a test script.

freshtonic commented on lib/graphql/lang/ast_reducer.ex:
```diff
@@ -0,0 +1,99 @@
+  defp visit(context, [child|rest], visitor, acc) do
+    { next_action, acc } = visit(context, child, visitor, acc)
+    case next_action do
+      :continue -> visit(context, rest, visitor, acc)
```
Calling visit recursively like this means I actually _will_ need to pop the stack in the context. Derp.

aweiker commented on lib/graphql/lang/ast_reducer.ex:
```diff
@@ -0,0 +1,105 @@
+
+# Used for computing a result as a function of the an AST traversal.  The
```
I'm curious, I didn't find anything named _visitor_ when [searching in the elixir code](https://github.com/elixir-lang/elixir/search?utf8=%E2%9C%93&q=visitor&type=Code) so I wonder how they solved this problem.

freshtonic replied:
It's certainly an interesting question. I haven't looked yet, but I am interested.

joshprice replied:
Please use `@moduledoc` for this

freshtonic replied:
@joshprice 👍

freshtonic commented on lib/graphql/type/definition.ex:
```diff
@@ -28,4 +27,17 @@ defmodule GraphQL.Type do
   defmodule NonNull do
     defstruct ofType: nil
   end
```
I expect this will need to change to accomodate @markolson's in-progress PR.

freshtonic committed:
Stack implementation for forthcoming AST Reducer

freshtonic committed:
Implement the AST Reducer

The AST Reducer walks the AST of a GraphQL query and invokes a Visitor
as nodes of the AST are traversed. The Visitor is responsible for
maintaining an accumulator.

At the end of the traversal, the final values in the accumulator can be
obtained. This lays the groundwork for implementing query validators
(the majority, if not all, will be Visitor implementations).

freshtonic committed:
Provide means to combine multiple Visitors into one

This will allow us to write a type information visitor that 'wraps' a
validation visitor so that when *enter* and *leave* are called on the
validation visitor, type information for the current node will be
available on the accumulator.

It also makes it easy to insert visitors in a chain and debug what's
going on when writing tests.

freshtonic committed:
Ensure that protocol implementations in tests are consolidated

freshtonic committed:
Implement the TypeInfoVisitor

This injects type information into the accumulator for subsequent
visitors (that have been combined using the CompositeVisitor) have
access to the type information of the current AST node.

NOTE: this is complex and there are no tests!

The implementation is translated from the JS version (which also has no
direct tests) but it tested indirectly via the validation tests. I plan
on taking the same approach.

joshprice commented on lib/graphql/lang/ast/composite_visitor.ex:
```diff
@@ -0,0 +1,108 @@
+
+defmodule GraphQL.Lang.AST.CompositeVisitor do
+  @moduledoc """
+  A CompositeVisitor composes two Visitor implementations into a single Visitor.
```
Well worth explaining inner v outer here as the distinction is non-obvious

joshprice commented on lib/graphql/lang/ast/reducer.ex:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  defp visit_children(node = %{kind: kind}, visitor, accumulator) when is_atom(kind) do
+    children = for child_key <- Nodes.kinds[node[:kind]], !!node[child_key], do: node[child_key]
```
`Map.has_key?`

joshprice commented on lib/graphql/lang/ast/composite_visitor.ex:
```diff
@@ -0,0 +1,108 @@
+  def compose(visitors) when is_list(visitors) and length(visitors) >= 2 do
+    [outer_visitor|[inner_visitor|rest]] = visitors 
+    composite_visitor = compose(outer_visitor, inner_visitor)
+    if length(rest) > 0 do
```
Make 2 extra fn head to handle 0 and 1 cases would be more obvious

joshprice commented on lib/graphql/lang/ast/composite_visitor.ex:
```diff
@@ -0,0 +1,108 @@
+
+  # Visits two visitors in the order specified and invokes the supplied Visitor function.
+  defp call_in_order(v1, v2, node, accumulator, fun) do
+    { v1_next_action, v1_accumulator } = fun.(v1, node, accumulator)
```
Remove excess padding see https://github.com/lexmag/elixir-style-guide#source-code-layout

joshprice commented on lib/graphql/lang/ast/composite_visitor.ex:
```diff
@@ -0,0 +1,108 @@
+  Returns the accumulator of the *inner* visitor.
+  """
+  def init(composite_visitor, accumulator) do
+    accumulator = composite_visitor.outer_visitor |> InitialisingVisitor.init(accumulator)
```
Remove pipeline as there is only a single function call https://github.com/lexmag/elixir-style-guide#needless-pipeline

joshprice commented on lib/graphql/lang/ast/type_info.ex:
```diff
@@ -0,0 +1,68 @@
+defmodule GraphQL.Lang.AST.TypeInfo do
+  @moduledoc ~S"""
+  TypeInfo maintains type metadata pertaining to the current node of a query AST,
+  and is used by the TypeInfoVistor.
```
"and generated by the TypeInfoVisitor"

joshprice commented on lib/graphql/lang/ast/type_info.ex:
```diff
@@ -0,0 +1,68 @@
+  """
+  def type(type_info), do: type_info.type_stack |> Stack.peek()
+
+  def named_type(type_info, type) do
```
pattern match?

joshprice commented:
:+1: looks good, formatting changes then :ship:

freshtonic committed:
Refactor: simplify &CompositeVisitor.compose/1

freshtonic committed:
Remove unnecessary piping

freshtonic committed:
Improved moduledoc for CompositeVisitor

freshtonic committed:
Minor refactor to improve code legibility

freshtonic committed:
Remove unused aliases

freshtonic committed:
Style: removed inner padding between "{" and "}"

freshtonic committed:
Docfix: moduledoc for TypeInfo

joshprice commented:
:+1:
